geode-core/src/main/java/org/apache/geode/cache/execute/Execution.java

```diff
@@ -51,7 +51,8 @@
    *         {@link FunctionService#onRegion(org.apache.geode.cache.Region)}
    * @since GemFire 6.0
    */
-  Execution<ArgumentT, ReturnT, AggregatorT> withFilter(Set<?> filter);
+  <ArgumentT, ReturnT, AggregatorT> Execution<ArgumentT, ReturnT, AggregatorT> withFilter(
```

Comment:
I don't understand why the type parameters are required at the beginning of this declaration—they are already defined on the class.

Reply:
@Bill they are required when you try to use them to chain things. We didn't have them originally and that is what broke your code.

Reply:
I kind of understand the effect. I don't know why it works though.

Do you have a reference that explains why it works? I'd love that for my mental toolkit.

I've seen situations where a complex expression had to be broken in two for the compiler to understand it, like this:

https://e.printstacktrace.blog/java-type-inference-generic-methods-chain-call/
https://github.com/Bill/FJ/blob/master/src/test/java/com/thoughtpropulsion/fj/ClosedTest.java#L36

@agingade found that without the changes in this PR, he was able to eliminate the compiler errors in `FunctionRetryTestBase.java` by changing e.g. this:

```java
      case REGION:
        execution =
            FunctionService.onRegion(clusterStartupRule.getClientCache().getRegion(regionName))
                .setArguments(200);
```

to this:

```java
      case REGION:
        final Execution<Integer, Long, List<Long>>
            tmp =
            FunctionService.onRegion(clusterStartupRule.getClientCache().getRegion(regionName));
        execution = tmp.setArguments(200);
        break;
```

Reply:
PR looks good to me. I'll educate myself on the details of _why_ this fix works, on my own time after the PR is merged.

Reply:
@Bill We fixed the issues earlier by breaking things up but when looking at it with @jinmeiliao today we found that the approach we take with `static` methods works on `instance` methods as well.

In that case the `Type` parameters you put before the return type tell the compiler to look at the receiver of the return object for type parameters and infer them from there.

Reply:
Thanks @jmelchio I'll have to ponder that. Here is some closely related info: https://blog.jooq.org/2013/11/25/a-lesser-known-java-8-feature-generalized-target-type-inference/

Reply:
This PR introduces a new bug:

https://issues.apache.org/jira/browse/GEODE-6884

Reply:
Yes, more experiments have shown that there are holes when try to chain. Need to go back to drawing board.